File: src/main/java/com/bajdcc/LALR1/syntax/exp/PropertyExp.java
package com.bajdcc.LALR1.syntax.exp;

import com.bajdcc.LALR1.semantic.token.ISemanticAction;
import com.bajdcc.LALR1.syntax.ISyntaxComponent;
import com.bajdcc.LALR1.syntax.ISyntaxComponentVisitor;
import com.bajdcc.LALR1.syntax.handler.IErrorHandler;
import com.bajdcc.util.VisitBag;

/**
 * 文法规则（属性）
 *
 * @author bajdcc
 */
public class PropertyExp implements ISyntaxComponent {

	/**
	 * 子表达式
	 */
	public ISyntaxComponent expression = null;

	/**
	 * 存储序号
	 */
	public int iStorage;

	/**
	 * 子表达式
	 */
	public IErrorHandler errorHandler;

	/**
	 * 动作名称
	 */
	public ISemanticAction actionHandler = null;

	public PropertyExp(int id, IErrorHandler handler) {
		iStorage = id;
		errorHandler = handler;
	}

	@Override
	public void visit(ISyntaxComponentVisitor visitor) {
		VisitBag bag = new VisitBag();
		visitor.visitBegin(this, bag);
		if (bag.getVisitChildren()) {
			expression.visit(visitor);
		}
		if (bag.getVisitEnd()) {
			visitor.visitEnd(this);
		}
	}
}

File: src/main/java/com/bajdcc/LALR1/syntax/exp/PropertyExp.kt
package com.bajdcc.LALR1.syntax.exp

import com.bajdcc.LALR1.semantic.token.ISemanticAction
import com.bajdcc.LALR1.syntax.ISyntaxComponent
import com.bajdcc.LALR1.syntax.ISyntaxComponentVisitor
import com.bajdcc.LALR1.syntax.handler.IErrorHandler
import com.bajdcc.util.VisitBag

/**
 * 文法规则（属性）
 * [iStorage] 存储序号
 * [expression] 子表达式
 * @author bajdcc
 */
class PropertyExp(var iStorage: Int,
                  var expression: ISyntaxComponent) : ISyntaxComponent {

    /**
     * 错误处理
     */
    var errorHandler: IErrorHandler? = null

    /**
     * 动作名称
     */
    var actionHandler: ISemanticAction? = null

    override fun visit(visitor: ISyntaxComponentVisitor) {
        val bag = VisitBag()
        visitor.visitBegin(this, bag)
        if (bag.visitChildren) {
            expression.visit(visitor)
        }
        if (bag.visitEnd) {
            visitor.visitEnd(this)
        }
    }
}
